### PR TITLE
Implement saving and loading camera settings

### PR DESCRIPTION
**Pull request by seanharmer**
Allows to easily get reproducable cameras between runs to be able
to compare images.

### CODE REVIEW

File: tools/assetpipelineeditor/camerawidget.cpp [hunk 1]
```diff
@@ -30,6 +30,33 @@
...
+#include <QJsonObject>
+#include <QSettings>
+
+namespace {
+    const QLatin1String LASTPATHSETTING("mainwindow/lastPath");
+
+    QJsonObject toJson(const QVector3D &v)
+    {
+        QJsonObject json;
+        json["x"] = static_cast<double>(v.x());
```
Comment (lemirep) [inline tools/assetpipelineeditor/camerawidget.cpp]: is "x" a QString or QLatin1String here ?

File: tools/assetpipelineeditor/camerawidget.cpp [hunk 2]
```diff
@@ -227,6 +256,61 @@ void CameraWidget::setFoV(double fov)
...
+    settings.setValue(LASTPATHSETTING, fi.absolutePath());
+
+    QFile saveFile(filename);
+    if (!saveFile.open(QIODevice::WriteOnly)) {
+        qWarning("Couldn't open camera settings save file.");
+        return;
+    }
+
+    QJsonObject cameraObject;
+    write(cameraObject);
```
Comment (lemirep) [inline tools/assetpipelineeditor/camerawidget.cpp]: write/read are a bit confusing as they are usually related to either the system calls or QIODevice read/write. How about doing QJsonObject cameraObject = toJson(); and fromJson() to restore?

File: tools/assetpipelineeditor/camerawidget.cpp [hunk 3]
```diff
@@ -227,6 +256,61 @@ void CameraWidget::setFoV(double fov)
...
+    QFileInfo fi(filename);
+    settings.setValue(LASTPATHSETTING, fi.absolutePath());
+
+    QFile loadFile(filename);
+    if (!loadFile.open(QIODevice::ReadOnly)) {
+        qWarning("Couldn't open camera settings load file.");
+        return;
+    }
+
+    QByteArray savedData = loadFile.readAll();
```
Comment (lemirep) [inline tools/assetpipelineeditor/camerawidget.cpp]: const QByteArray

File: tools/assetpipelineeditor/camerawidget.cpp [hunk 4]
```diff
@@ -227,6 +256,61 @@ void CameraWidget::setFoV(double fov)
...
+    QJsonDocument loadDoc(QJsonDocument::fromJson(savedData));
```
Comment (lemirep) [inline tools/assetpipelineeditor/camerawidget.cpp]: const QJsonDocument

File: tools/assetpipelineeditor/camerawidget.cpp [hunk 5]
```diff
@@ -30,6 +30,33 @@
...
+#include <QFileDialog>
+#include <QFileInfo>
+#include <QJsonDocument>
+#include <QJsonObject>
+#include <QSettings>
+
+namespace {
+    const QLatin1String LASTPATHSETTING("mainwindow/lastPath");
+
+    QJsonObject toJson(const QVector3D &v)
```
Comment (lemirep) [inline tools/assetpipelineeditor/camerawidget.cpp]: vector3DToJson() ? or you need to have a templated T fromJson(const QJsonObject) below

File: tools/assetpipelineeditor/camerawidget.cpp [hunk 6]
```diff
@@ -257,3 +341,55 @@ void CameraWidget::updateWidgetValues()
...
+    json["viewCenter"] = toJson(m_camera->viewCenter());
+    json["upVector"] = toJson(m_camera->upVector());
+    json["nearPlane"] = static_cast<double>(m_camera->nearPlane());
+    json["farPlane"] = static_cast<double>(m_camera->farPlane());
+
+    const bool isPerspective = m_camera->projectionType() == Qt3DRender::QCameraLens::PerspectiveProjection;
+    if (isPerspective) {
+        json["projectionType"] = QLatin1String("Perspective");
+        json["fieldOfView"] = static_cast<double>(m_camera->fieldOfView());
+        json["aspectRatio"] = static_cast<double>(m_camera->aspectRatio());
```
Comment (TimoBuskeKdab) [inline tools/assetpipelineeditor/camerawidget.cpp]: is this something we want to save and we can load properly?
it doesn't really hurt but may lead to distorted output after loading in a window with different aspect until aspect changes...

Comment (seanharmer) [reply tools/assetpipelineeditor/camerawidget.cpp]: Hmm yeah. Let's get it in and see how it behaves - perhaps we can add a follow up to trigger a new fov and aspect ratio or left/right/top/bottom calc to be done somehow.

**Pull request merged**